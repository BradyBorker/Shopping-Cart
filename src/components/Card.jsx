import { useState, useContext } from "react"
import { CartContext } from "../App";
import PropTypes from 'prop-types';
import styles from "../styles/card.module.css"

const Card = ({ product }) => {
    const [counter, setCounter] = useState(0);
    const setCart = useContext(CartContext)

    const updateQuantity = (e) => {
        if (e.target.value.length > 1 && e.target.value[0] === '0') {
            console.log('hi')
            return setCounter(e.target.value.substring(1))
        }

        (e.target.value <= 0) ? setCounter(0) : setCounter(e.target.value)
    }

    const addToCart = (e) => {
        e.preventDefault();
        setCounter(0);

        if (Number(counter) !== 0) {
            setCart((cart) => {
                if (cart[product.id] === undefined) {
                    return {
                        ...cart, [`${product.id}`]: {
                            product,
                            count: Number(counter)
                        }
                    }
                };
                return {
                    ...cart, [`${[product.id]}`]: {
                        product,
                        count: Number(counter) + cart[product.id].count
                    }
                };
            })

            const input = e.target['0']
            input.classList.toggle(styles.goodSubmit);
            setTimeout(() => {
                input.classList.toggle(styles.goodSubmit);
            }, 1000)
        }
    }

    return (
        <div className={styles.card}>
            <img className={styles.image} src={product.image} alt={product.description} />
            <div className={styles.title}>{product.title}</div>
            <div className={styles.price}>{`$${product.price}`}</div>
            <form className={styles.addToCart} onSubmit={(e) => addToCart(e)} aria-label="Add to cart">
                <div className={styles.counter}>
                    <input type="number" value={counter} onChange={(e) => updateQuantity(e)} />
                </div>
                <button className={styles.button} type="submit">Add to Cart</button>
            </form>
        </div>
    )
}

Card.propTypes = {
    product: PropTypes.object.isRequired,
}

export default Card