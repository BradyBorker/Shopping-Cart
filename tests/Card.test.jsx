import { render, screen } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import { BrowserRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import Card from '../src/components/Card';

const product = { id: 1, title: 'product1', image: '', price: 1, description: 'product1 description' }

describe('Card component', () => {
    it('renders the Card component', () => {
        const { container } = render(<BrowserRouter><Card product={product} /></BrowserRouter>);
        expect(container).toMatchSnapshot();
    })

    it('renders the products information', () => {
        render(<BrowserRouter><Card product={product} /></BrowserRouter>);

        const img = screen.getByAltText('product1 description');
        expect(img).toBeInTheDocument();

        const title = screen.getByText('product1');
        expect(title).toBeInTheDocument();

        const price = screen.getByText('$1');
        expect(price).toBeInTheDocument();
    })

    it('renders the products form', () => {
        render(<BrowserRouter><Card product={product} /></BrowserRouter>);

        const form = screen.getByRole('form', { name: 'Add to cart' });
        expect(form).toBeInTheDocument();

        const input = screen.getByRole('spinbutton');
        expect(input).toBeInTheDocument();

        const button = screen.getByRole('button', { name: 'Add to Cart' })
        expect(button).toBeInTheDocument();
    })

    it('increments the inputs counter', async () => {
        const user = userEvent.setup();

        render(<BrowserRouter><Card product={product} /></BrowserRouter>);

        const input = screen.getByRole('spinbutton');
        await user.type(input, '1');

        expect(input.value).toBe('1');
    })

    it('prevents the inputs value from going below 0', async () => {
        const user = userEvent.setup();

        render(<BrowserRouter><Card product={product} /></BrowserRouter>);

        const input = screen.getByRole('spinbutton');
        await user.type(input, '-1');

        expect(input.value).toBe('1');
    })

    it('calls addToCart when button is clicked', async () => {
        const user = userEvent.setup();
        const addToCart = vi.fn();

        render(<BrowserRouter><Card product={product} /></BrowserRouter>);

        const button = screen.getByRole('button', { name: 'Add to Cart' })
        const form = screen.getByRole('form', { name: 'Add to cart' });
        form.onsubmit = addToCart;

        await user.click(button);

        expect(addToCart).toBeCalledTimes(1);
    })
})