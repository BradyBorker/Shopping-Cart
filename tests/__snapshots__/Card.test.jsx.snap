// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card component > renders the Card component 1`] = `
<div>
  <div
    class="_card_277467"
  >
    <img
      alt="product1 description"
      class="_image_277467"
      src=""
    />
    <div
      class="_title_277467"
    >
      product1
    </div>
    <div
      class="_price_277467"
    >
      $1
    </div>
    <form
      aria-label="Add to cart"
      class="_addToCart_277467"
    >
      <div
        class="_counter_277467"
      >
        <input
          type="number"
          value="0"
        />
      </div>
      <button
        class="_button_277467"
        type="submit"
      >
        Add to Cart
      </button>
    </form>
  </div>
</div>
`;
